sdk/ml/azure-ai-ml/tests/dsl/e2etests/test_dsl_fl.py

```diff
@@ -207,48 +241,39 @@ def silo_step_func(
         aggregation_datastore = datastores[3].name
 
         # Define the fl pipeline
-        @pipeline()
-        def fl_pipeline():
-            fl_node = fl_scatter_gather(
-                silo_configs=silo_configs,
-                silo_component=silo_step_func,
-                aggregation_component=aggregation_step,
-                aggregation_compute=aggregation_compute,
-                aggregation_datastore=aggregation_datastore,
-                shared_silo_kwargs=silo_kwargs,
-                aggregation_kwargs=agg_kwargs,
-                silo_to_aggregation_argument_map=silo_to_aggregation_argument_map,
-                aggregation_to_silo_argument_map=aggregation_to_silo_argument_map,
-                max_iterations=iterations,
-            )
-            return {"pipeline_output": fl_node.outputs["aggregated_output"]}
+        fl_node = fl_scatter_gather(
```

Comment:
This change to the test confuses me. The purpose of an E2E test is to test intended behavior, including API interactions, as closely as possible. Why are we no longer using the node in its intended way by placing the node inside a @pipeline decorated function? Has the expected use case changed to assume that it's a standalone piece? That's not how DSL nodes are intended to be used - it might as well be its own annotation in that case, which we specifically chose to avoid a while ago due to implementation challenges.

Reply:
The do_while loop needs to be in the @pipeline decorator method. Therefore, I had to place the scatter-gather-body in the same method. This resulted in returning the complete pipeline graph that is needed for an FL graph. Hence, I assigned the resulting Pipeline component to one of the Scatter-Gather node’s attributes and directly accessed it here to create a PipelineJob. Furthermore, this does not imply that we cannot use Scatter-gather node in a @pipeline decorated function, but that would lead to a pipeline containing a single pipeline component and then the actual scatter-gather node. With regards to API interaction, we can only go one layer deep to access pipeline component’s jobs, so unfortunately, we cannot validate the behavior of inner components. Please correct me if I'm missing anything.

Reply:
I think it would be pertinent to eventually add another e2e test that executes the expected user behavior (running this node inside a @pipeline-decorated function), but given the fact that e2e tests for this package are broken, that's not a priority. Just create an issue, and comment the created issue number in this file to keep track of it please.

Reply:
https://github.com/Azure/azure-sdk-for-python/issues/29680